protoc-gen-swagger/defs.bzl

```diff
@@ -0,0 +1,79 @@
+def _collect_includes(srcs):
+    includes = ["."]
+    for src in srcs:
+        include = src.dirname
+        if include and not include in includes:
+            includes += [include]
+
+    return includes
+
+def _run_proto_gen_swagger(proto_service, transitive_proto_srcs, actions, protoc, protoc_gen_swagger):
+    swagger_file = actions.declare_file(
+        "%s.swagger.json" % proto_service.basename[:-len(".proto")],
+        sibling = proto_service,
+    )
+
+    args = actions.args()
+    args.add("--plugin=%s" % protoc_gen_swagger.path)
+    args.add("--swagger_out=logtostderr=true:%s" % swagger_file.dirname)
+    args.add("-Iexternal/com_google_protobuf/src")
+    args.add("-Iexternal/grpc_ecosystem_grpc_gateway/third_party/googleapis/")
+    args.add(["-I%s" % include for include in _collect_includes(transitive_proto_srcs)])
+    args.add(proto_service.basename)
+
+    actions.run(
+        executable = protoc,
+        inputs = transitive_proto_srcs + [protoc_gen_swagger],
+        outputs = [swagger_file],
+        arguments = [args],
+    )
+
+    return swagger_file
+
+def _proto_gen_swagger_impl(ctx):
+    transitive_proto_srcs = depset([ctx.file.proto_service] + ctx.files._googleapis + ctx.files._well_known_protos)
+    for dep in ctx.attr.deps:
+        transitive_proto_srcs = depset(transitive=[transitive_proto_srcs, dep.proto.transitive_sources])
+
+    return struct(
+        files=depset([
+            _run_proto_gen_swagger(
+                ctx.file.proto_service,
+                transitive_proto_srcs.to_list(),
+                ctx.actions,
+                ctx.executable._protoc,
+                ctx.executable._protoc_gen_swagger,
+            )
+        ])
+    )
+
+protoc_gen_swagger = rule(
+    attrs = {
+        "proto_service": attr.label(
```

Comment:
Instead of taking the `.proto` file, I think it would be better to take a `proto_library` so it is consistent on how the `go_proto_library` works. Take a look at the implementation [here](https://github.com/bazelbuild/rules_go/blob/master/proto/go_proto_library.bzl).

Reply:
Donezo

Reply:
You could rename to `proto` instead of `proto_service`.